src/scaffoldmaker/meshtypes/meshtype_3d_lung2.py

```diff
@@ -0,0 +1,2139 @@
+'''
+Generates a 3D generic lung mesh.
+'''
+import copy
+import math
+from scaffoldmaker.utils.interpolation import sampleCubicHermiteCurves, interpolateSampleCubicHermite, \
+    smoothCubicHermiteDerivativesLine, interpolateSampleLinear
+from scaffoldmaker.annotation.annotationgroup import AnnotationGroup, findOrCreateAnnotationGroupForTerm, getAnnotationGroupForTerm
+from scaffoldmaker.annotation.lung_terms import get_lung_term
+from scaffoldmaker.meshtypes.scaffold_base import Scaffold_base
+from scaffoldmaker.utils.eft_utils import remapEftLocalNodes, remapEftNodeValueLabel, remapEftNodeValueLabelsVersion, setEftScaleFactorIds
+from scaffoldmaker.utils.cylindermesh import createEllipsePerimeter
+from scaffoldmaker.utils.eftfactory_tricubichermite import eftfactory_tricubichermite
+from opencmiss.utils.zinc.general import ChangeManager
+from scaffoldmaker.utils.geometry import createEllipsoidPoints, getEllipseRadiansToX, getEllipseArcLength, getApproximateEllipsePerimeter, \
+    updateEllipseAngleByArcLength
+from scaffoldmaker.utils.interpolation import DerivativeScalingMode
+from scaffoldmaker.utils.derivativemoothing import DerivativeSmoothing
+from scaffoldmaker.utils.meshrefinement import MeshRefinement
+from scaffoldmaker.utils.vector import magnitude, setMagnitude, crossproduct3, normalise
+from opencmiss.utils.zinc.field import Field, findOrCreateFieldCoordinates, findOrCreateFieldGroup, \
+    findOrCreateFieldNodeGroup, findOrCreateFieldStoredMeshLocation, findOrCreateFieldStoredString
+from opencmiss.utils.zinc.finiteelement import get_element_node_identifiers
+from opencmiss.zinc.element import Element
+from opencmiss.zinc.node import Node
+
+class MeshType_3d_lung2(Scaffold_base):
+    '''
+    Generic 3D lung scaffold.
+    '''
+
+    """
+        Generates an ellipsoid with a tear-shaped base for the lung mathematically,
+         with x, y, z length and the position, angle of the oblique fissure.
+         Regions and markers of the lung are annotated.
+    """
+
+    @staticmethod
+    def getName():
+        return '3D Lung 2'
+
+    @staticmethod
+    def getParameterSetNames():
+        return [
+            'Default',
+            'Human 1',
+            'Mouse 1',
+            'Rat 1',
+            'Pig 1',
+            'Material'
+        ]
+
+    @classmethod
+    def getDefaultOptions(cls, parameterSetName='Default'):
+        if parameterSetName == 'Default':
+            parameterSetName = 'Human 1'
+        options = {
+            'Base parameter set': parameterSetName,
+            'Number of left lung lobes': 2,
+            'Left-right lung spacing': 1.0,
+            'Left-right apex closing space': 0.0,
+            'Left-right apex along y-axis': 0.0,
+            'Left lung width': 0.25,
+            'Left lung depth': 0.5,
```

Comment:
Left and Right lung width, depth should be twice this. Users understand whole dimensions like diameters, but we coders should convert them to radius or major/minor axis lengths by dividing by 2 when we use them.
Height is fine since we don’t include the bottom half of the ellipsoid.